File: macos/Sources/OpenShrimpCore/ControlClient.swift
import Foundation
import Network

/// Client for the core's control channel (a Unix socket on macOS).
///
/// The core pushes unsolicited event frames on the same connection as RPC
/// replies, so the read loop lives in one place and hands responses back to
/// waiting callers by id.  A caller that read the socket directly would sooner
/// or later consume an event as if it were its own reply.
///
/// Being an actor is what serialises the callers.  The watchdog poll, the
/// event-driven refresh and an explicit stop can all call in at once, and
/// interleaved writes would produce a line the server cannot parse — both
/// callers would then wait out their timeout, which the supervisor reads as
/// "the core refused to stop" and escalates to a kill.
actor ControlClient {
    /// How long a request waits for its reply.
    static let replyTimeout: TimeInterval = 15

    /// How long one connect attempt is given.  Short because a core that is
    /// listening answers immediately, and a missing socket is the common case.
    static let connectTimeout: TimeInterval = 1

    /// Delay between reconnect attempts.
    static let retryDelay: TimeInterval = 0.5

    /// `sun_path` in `struct sockaddr_un` is 104 bytes on macOS, one of which
    /// is the terminating NUL.  The server refuses to bind past this with a
    /// readable message; the client checks the same limit so that an address
    /// nothing can ever listen on is reported as such rather than as a core
    /// that is merely not running.
    static let sunPathMax = 103

    /// One serial queue for every connection's callbacks, which is what makes
    /// the one-shot resume guards below safe without a lock.
    private static let queue = DispatchQueue(label: "openshrimp.control")

    let endpoint: String

    private var connection: NWConnection?
    private var readLoop: Task<Void, Never>?
    private var disposed = false

    /// Distinguishes a drop reported by the current read loop from one
    /// reported by a loop whose connection has already been replaced.  Without
    /// it, the old loop unwinding after a successful reconnect would announce
    /// a disconnection that has already been recovered from.
    private var generation = 0

    /// Monotonic from 1 for the life of the client, and deliberately not reset
    /// across a reconnect: a reply to a request issued before the core
    /// re-execed must never be mistaken for a reply to one issued after it.
    private var nextId = 0

    /// A reply can arrive while its caller is still suspended in the send that
    /// asked for it, so a waiter has to be able to hold an answer that nobody
    /// is waiting on yet.  The first of reply and timeout wins.
    private enum Waiter {
        case waiting(CheckedContinuation<ControlFrame?, Never>)
        case answered(ControlFrame?)
    }

    private var waiters: [Int: Waiter] = [:]

    private var onEvent: (@Sendable (String) -> Void)?
    private var onDisconnect: (@Sendable () -> Void)?

    private(set) var isConnected = false

    init(instanceName: String?) {
        endpoint = CorePaths.controlSocket(instanceName: instanceName)
    }

    /// True when no socket could ever be bound at this address, whatever the
    /// core does.  Only the instance name is under anyone's control.
    nonisolated var endpointExceedsSunPath: Bool {
        endpoint.utf8.count > Self.sunPathMax
    }

    func setHandlers(
        onEvent: (@Sendable (String) -> Void)? = nil,
        onDisconnect: (@Sendable () -> Void)? = nil
    ) {
        self.onEvent = onEvent
        self.onDisconnect = onDisconnect
    }

    // -- Connection ----------------------------------------------------------

    /// Attempt one connection.  False means "not running", including when the
    /// socket file is absent: the runtime directory is subject to periodic OS
    /// cleanup, and the server re-creates and re-binds it on every start, so a
    /// purge while the core is stopped is not a fault to report.
    func connect(timeout: TimeInterval = ControlClient.connectTimeout) async -> Bool {
        guard !disposed, !endpointExceedsSunPath else { return false }
        teardown()

        let connection = NWConnection(to: .unix(path: endpoint), using: .tcp)
        let ready = await Self.start(connection, timeout: timeout)
        guard ready, !disposed else {
            connection.cancel()
            return false
        }

        generation += 1
        let generation = self.generation
        self.connection = connection
        isConnected = true
        readLoop = Task { [weak self] in
            await self?.runReadLoop(connection, generation: generation)
        }
        return true
    }

    /// Reconnect across a core re-exec.  `/restart` and auto-update replace the
    /// process, so the pid changes and the socket drops; the endpoint path is
    /// stable, which is why liveness is judged by the endpoint and not by
    /// tracking a child process.
    func reconnect(within: TimeInterval) async -> Bool {
        let deadline = Date().addingTimeInterval(within)
        while Date() < deadline {
            // Bail the moment someone disposes us, or a reconnect that wins the
            // race would hand back a live connection and read loop that nothing
            // owns.
            if disposed { return false }
            if await connect() { return true }
            if disposed { return false }
            try? await Task.sleep(nanoseconds: UInt64(Self.retryDelay * 1_000_000_000))
        }
        return false
    }

    func dispose() {
        disposed = true
        teardown()
    }

    /// Drop the current connection, silently.
    ///
    /// A teardown we chose never announces itself: the disconnect handler is
    /// what starts a reconnect, so announcing one here would re-enter the very
    /// reconnect that asked for the teardown.  Only the read loop unwinding on
    /// its own reports a loss.
    private func teardown() {
        let connection = self.connection
        self.connection = nil
        readLoop?.cancel()
        readLoop = nil
        isConnected = false
        // Cancelling the connection is what unblocks a read parked in the loop;
        // cancelling the task alone would leave it suspended forever.
        connection?.cancel()
        failWaiters()
    }

    // -- Requests ------------------------------------------------------------

    func status() async -> CoreStatus? {
        await call("status")?.status
    }

    /// True when the core accepted the request, not when it has finished
    /// acting on it — the reply is written before the core unwinds.
    func shutdown() async -> Bool {
        await call("shutdown") != nil
    }

    func restart() async -> Bool {
        await call("restart") != nil
    }

    /// Never throws: every failure path answers nil, and the supervisor's state
    /// machine is written against that.
    private func call(_ method: String) async -> ControlFrame? {
        guard let connection, isConnected else { return nil }

        nextId += 1
        let id = nextId
        guard let payload = try? JSONSerialization.data(
            withJSONObject: ["id": id, "method": method]
        ) else { return nil }

        let timeout = Task { [weak self] in
            try? await Task.sleep(nanoseconds: UInt64(Self.replyTimeout * 1_000_000_000))
            guard !Task.isCancelled else { return }
            await self?.deliver(id: id, frame: nil)
        }
        defer { timeout.cancel() }

        // One send carries the whole frame including its terminator.
        // NWConnection queues content in call order, so no second caller can
        // interleave half a line into this one.
        guard await Self.send(payload + Data([0x0A]), over: connection) else {
            waiters[id] = nil
            return nil
        }
        return await awaitReply(id: id)
    }

    private func awaitReply(id: Int) async -> ControlFrame? {
        await withCheckedContinuation { continuation in
            if case .answered(let frame) = waiters[id] {
                waiters[id] = nil
                continuation.resume(returning: frame)
            } else {
                waiters[id] = .waiting(continuation)
            }
        }
    }

    private func deliver(id: Int, frame: ControlFrame?) {
        switch waiters[id] {
        case .waiting(let continuation):
            waiters[id] = nil
            continuation.resume(returning: frame)
        case .answered:
            break
        case nil:
            waiters[id] = .answered(frame)
        }
    }

    private func failWaiters() {
        for (id, waiter) in waiters {
            waiters[id] = nil
            if case .waiting(let continuation) = waiter {
                continuation.resume(returning: nil)
            }
        }
    }

    // -- Read loop -----------------------------------------------------------

    private func runReadLoop(_ connection: NWConnection, generation: Int) async {
        var buffer = Data()

        while let chunk = await Self.receive(connection) {
            buffer.append(chunk)
            while let newline = buffer.firstIndex(of: 0x0A) {
                let line = Data(buffer[buffer.startIndex..<newline])
                buffer.removeSubrange(buffer.startIndex...newline)
                guard !line.isEmpty, let frame = ControlFrame(line: line) else { continue }
                dispatch(frame)
            }
        }

        // Only the loop that owns the live connection may report its loss.
        guard generation == self.generation, isConnected else { return }
        isConnected = false
        self.connection = nil
        failWaiters()
        onDisconnect?()
    }

    private func dispatch(_ frame: ControlFrame) {
        if let event = frame.event {
            onEvent?(event)
            return
        }
        guard let id = frame.id else { return }
        deliver(id: id, frame: frame)
    }

    // -- Transport -----------------------------------------------------------

    /// Resumes its continuation exactly once.  Safe without a lock because
    /// every path that touches it runs on `queue`, which is serial.
    private final class OneShot: @unchecked Sendable {
        private var fired = false
        func fire(_ body: () -> Void) {
            guard !fired else { return }
            fired = true
            body()
        }
    }

    private static func start(_ connection: NWConnection, timeout: TimeInterval) async -> Bool {
        await withCheckedContinuation { continuation in
            let once = OneShot()
            connection.stateUpdateHandler = { state in
                switch state {
                case .ready:
                    once.fire { continuation.resume(returning: true) }
                case .waiting, .failed, .cancelled:
                    // .waiting covers a socket file that is absent or that
                    // nothing is accepting on.  NWConnection would retry it
                    // forever; one attempt is what the caller asked for.
                    once.fire { continuation.resume(returning: false) }
                default:
                    break
                }
            }
            connection.start(queue: queue)
            queue.asyncAfter(deadline: .now() + timeout) {
                once.fire { continuation.resume(returning: false) }
            }
        }
    }

    private static func send(_ data: Data, over connection: NWConnection) async -> Bool {
        await withCheckedContinuation { continuation in
            connection.send(content: data, completion: .contentProcessed { error in
                continuation.resume(returning: error == nil)
            })
        }
    }

    /// One read.  Nil ends the loop, which is the only signal the connection is
    /// gone — the frame cap is the server's concern, so no ceiling is imposed
    /// on what may accumulate here.
    private static func receive(_ connection: NWConnection) async -> Data? {
        await withCheckedContinuation { continuation in
            connection.receive(minimumIncompleteLength: 1, maximumLength: 64 * 1024) {
                data, _, _, _ in
                // Having asked for a minimum of one byte, an empty completion
                // is the end of the stream — or a state the framework does not
                // document, which is treated the same way.  Looping on it
                // instead would spin a core; ending early costs one reconnect
                // inside the grace window.
                continuation.resume(returning: (data?.isEmpty == false) ? data : nil)
            }
        }
    }
}

File: windows/src/OpenShrimp.Tray/Core/ControlClient.cs
using System.IO.Pipes;
using System.Text;
using System.Text.Json;

namespace OpenShrimp.Tray.Core;

/// <summary>
/// Client for the core's control channel (a named pipe on Windows).
///
/// The core pushes unsolicited event frames on the same connection as RPC
/// replies, so the read loop lives in one place and hands responses back to
/// waiting callers by id. A caller that read the socket directly would sooner
/// or later consume an event as if it were its own reply.
/// </summary>
internal sealed class ControlClient : IAsyncDisposable
{
    private readonly string _pipeName;
    private NamedPipeClientStream? _pipe;
    private StreamReader? _reader;
    private StreamWriter? _writer;
    private Task? _readLoop;
    private CancellationTokenSource? _cts;
    private volatile bool _disposed;

    private int _nextId;
    private readonly object _gate = new();
    private readonly Dictionary<int, TaskCompletionSource<ControlResponse>> _pending = new();

    // One frame at a time on the wire. The watchdog poll, the event-driven
    // refresh and an explicit stop can all call in at once; interleaved writes
    // would produce a line the server cannot parse, and both callers would then
    // wait out their timeout — which the supervisor reads as "the core refused
    // to stop" and escalates to a kill.
    private readonly SemaphoreSlim _writeLock = new(1, 1);

    // Guards connection setup and teardown against each other: a reconnect
    // driven by the read loop can otherwise race a Dispose from the UI.
    private readonly SemaphoreSlim _connectionLock = new(1, 1);

    /// <summary>Raised for every unsolicited event frame (state, stopping).</summary>
    public event Action<string, JsonElement?>? EventReceived;

    public event Action? Disconnected;

    /// <summary>
    /// Set once the read loop has ended, which is the first moment the far end
    /// going away is observable. <see cref="PipeStream.IsConnected"/> alone is
    /// a cached flag that a peer's death does not clear — it reported a live
    /// channel to a core that had already exited — so liveness is the two
    /// together.
    /// </summary>
    private volatile bool _closed;

    public bool IsConnected => !_closed && _pipe?.IsConnected == true;

    public ControlClient(string? instanceName)
    {
        // Mirrors control/server.py endpoint_address(): derived from the
        // instance name alone, so no discovery file is needed.
        var name = string.IsNullOrEmpty(instanceName) ? "openshrimp" : $"openshrimp-{instanceName}";
        _pipeName = $"{name}-control";
    }

    public async Task<bool> TryConnectAsync(TimeSpan timeout, CancellationToken ct = default)
    {
        var pipe = new NamedPipeClientStream(".", _pipeName, PipeDirection.InOut, PipeOptions.Asynchronous);
        try
        {
            await pipe.ConnectAsync((int)timeout.TotalMilliseconds, ct).ConfigureAwait(false);
        }
        catch (Exception) when (!ct.IsCancellationRequested)
        {
            await pipe.DisposeAsync().ConfigureAwait(false);
            return false;
        }

        _pipe = pipe;
        _closed = false;
        _reader = new StreamReader(pipe, new UTF8Encoding(false), false, 8192, leaveOpen: true);
        _writer = new StreamWriter(pipe, new UTF8Encoding(false), 8192, leaveOpen: true) { AutoFlush = false };
        _cts = new CancellationTokenSource();
        _readLoop = Task.Run(() => ReadLoopAsync(_cts.Token));
        return true;
    }

    /// <summary>
    /// Reconnects across a core re-exec. /restart and auto-update replace the
    /// process, so the pid changes and the pipe drops; the endpoint name is
    /// stable, which is why liveness is judged by the pipe and not by tracking
    /// a child process.
    /// </summary>
    public async Task<bool> ReconnectAsync(TimeSpan within, CancellationToken ct = default)
    {
        var deadline = DateTime.UtcNow + within;
        while (DateTime.UtcNow < deadline && !ct.IsCancellationRequested)
        {
            // Bail the moment someone disposes us, or a reconnect that wins the
            // race would hand back a live pipe and read loop that nothing owns.
            if (_disposed) return false;

            await DisconnectAsync().ConfigureAwait(false);
            if (_disposed) return false;
            if (await TryConnectAsync(TimeSpan.FromSeconds(1), ct).ConfigureAwait(false))
                return true;
            await Task.Delay(500, ct).ConfigureAwait(false);
        }
        return false;
    }

    public async Task<CoreStatus?> GetStatusAsync(CancellationToken ct = default)
    {
        var reply = await CallAsync("status", ct).ConfigureAwait(false);
        if (reply?.Result is null) return null;
        return reply.Result.Value.Deserialize<CoreStatus>(ControlJson.Options);
    }

    public Task<ControlResponse?> ShutdownAsync(CancellationToken ct = default) => CallAsync("shutdown", ct);

    public Task<ControlResponse?> RestartAsync(CancellationToken ct = default) => CallAsync("restart", ct);

    private async Task<ControlResponse?> CallAsync(string method, CancellationToken ct)
    {
        var writer = _writer;
        if (writer is null || !IsConnected) return null;

        int id;
        var tcs = new TaskCompletionSource<ControlResponse>(TaskCreationOptions.RunContinuationsAsynchronously);
        lock (_gate)
        {
            id = ++_nextId;
            _pending[id] = tcs;
        }

        // Serialise the frame and its terminator as one write under the lock:
        // two awaited writes could interleave with another caller's.
        var frame = JsonSerializer.Serialize(new { id, method }) + "\n";
        await _writeLock.WaitAsync(ct).ConfigureAwait(false);
        try
        {
            await writer.WriteAsync(frame).ConfigureAwait(false);
            await writer.FlushAsync(ct).ConfigureAwait(false);
        }
        catch (Exception)
        {
            lock (_gate) _pending.Remove(id);
            return null;
        }
        finally
        {
            _writeLock.Release();
        }

        using var timeout = CancellationTokenSource.CreateLinkedTokenSource(ct);
        timeout.CancelAfter(TimeSpan.FromSeconds(15));
        try
        {
            return await tcs.Task.WaitAsync(timeout.Token).ConfigureAwait(false);
        }
        catch (OperationCanceledException)
        {
            lock (_gate) _pending.Remove(id);
            return null;
        }
    }

    private async Task ReadLoopAsync(CancellationToken ct)
    {
        var reader = _reader;
        if (reader is null) return;

        try
        {
            while (!ct.IsCancellationRequested)
            {
                var line = await reader.ReadLineAsync(ct).ConfigureAwait(false);
                if (line is null) break;
                if (line.Length == 0) continue;

                ControlResponse? frame;
                try
                {
                    frame = JsonSerializer.Deserialize<ControlResponse>(line, ControlJson.Options);
                }
                catch (JsonException)
                {
                    continue;
                }
                if (frame is null) continue;

                if (frame.IsEvent)
                {
                    EventReceived?.Invoke(frame.Event!, frame.Data);
                    continue;
                }

                if (frame.Id is null || frame.Id.Value.ValueKind != JsonValueKind.Number) continue;
                var id = frame.Id.Value.GetInt32();
                TaskCompletionSource<ControlResponse>? waiter;
                lock (_gate)
                {
                    _pending.Remove(id, out waiter);
                }
                waiter?.TrySetResult(frame);
            }
        }
        catch (Exception)
        {
        }
        finally
        {
            _closed = true;
            FailPending();
            Disconnected?.Invoke();
        }
    }

    private void FailPending()
    {
        lock (_gate)
        {
            foreach (var waiter in _pending.Values)
                waiter.TrySetCanceled();
            _pending.Clear();
        }
    }

    private async Task DisconnectAsync()
    {
        await _connectionLock.WaitAsync().ConfigureAwait(false);
        try
        {
            var cts = _cts;
            var readLoop = _readLoop;
            _cts = null;
            _readLoop = null;

            cts?.Cancel();
            if (readLoop is not null)
            {
                try { await readLoop.ConfigureAwait(false); } catch { /* shutting down */ }
            }
            cts?.Dispose();

            // Letting go of a pipe whose other end has died must not throw:
            // disposing the writer flushes it, and flushing a broken pipe
            // raises. That happens on the ordinary path at the end of a
            // session, where the core is gone before the tray is asked to stop
            // it, and the exception would otherwise surface as a failed stop.
            try
            {
                _reader?.Dispose();
                _writer?.Dispose();
                if (_pipe is not null) await _pipe.DisposeAsync().ConfigureAwait(false);
            }
            catch (IOException) { /* the far end is already gone */ }
            catch (ObjectDisposedException) { }
            _reader = null;
            _writer = null;
            _pipe = null;
        }
        finally
        {
            _connectionLock.Release();
        }
    }

    public async ValueTask DisposeAsync()
    {
        _disposed = true;
        await DisconnectAsync().ConfigureAwait(false);
    }
}
